### ISSUE TITLE
connecting to CAS using a Bearer OAuth token over REST in full deployment (no user SHELL access)

### ISSUE THREAD

**Issue by GMoney1986** [sassoftware/python-swat#102]
Hi, could we get OAuth as an authorization option?  

I'm in a full SAS Viya deployment on RHEL where users do not have shell access or shell accounts on the workspace server and on the CAS controller.  As my organization has a tight lockdown on providing shell access to end users, our environment is set up so that the SASLogon service provides single sign on, and all connections to CAS are made through OAuth through the SAS Workspace context where a shared CAS service ID is used. 

If the swat.CAS connector could accept the OAuth token in Bearer format, or let me pass my user id, password, client id and client secret to generate the OAuth token to be used, that would allow me to use SWAT in my SAS Viya environment. 

Thank you

**Comment by kesmit13** [comment]
Actually, you can use OAuth tokens.  Just specify it as your password and leave the username blank.

**Comment by GMoney1986** [comment]
I think the issues I'm having must be related to the format of the OAuth token that I pass through to swat, or the type of token I'm getting generated from the SAS Viya REST API (using the /SASLogon/oauth/token#password endpoint on the SAS Viya REST API). I had assumed the Authorization to CAS would need to be in Bearer format based on the [SAS REST API documentation](https://developer.sas.com/apis/rest/Topics/#making-api-requests-with-access-tokens), which I can use when I manually connect to the CAS REST API, but doesn't seem to work when I follow the blank/None username and token in the password.

If I use the following, I get an authorization error with swat. 

Code
```
import os
import swat

os.environ['CAS_DEBUG_REQUESTS']="TRUE"
os.environ['CAS_DEBUG_REQUEST_BODIES']="TRUE"
os.environ['CAS_DEBUG_RESPONSES']="TRUE"
os.environ['CAS_TRACE_UI_ACTIONS']="TRUE"

conn = swat.CAS(sas_server_url+'/cas-shared-default-http/', password=oauth_token,  protocol='https')
```

Log:
```

PUT https://sas.corpserver.com:443/cas-shared-default-http/cas/sessions
User-Agent: python-requests/2.22.0
Accept-Encoding: gzip, deflate
Accept: */*
Connection: keep-alive
Content-Type: application/json
Content-Length: 0
Authorization: b'Basic aljdfklajsdfkjasdklfj='

{
  "error": "HTTP/1.1 401 Unauthorized",
  "code": "HTTP_401",
  "details": null,
  "disposition": null
}

---------------------------------------------------------------------------
SWATError                                 Traceback (most recent call last)
~\AppData\Roaming\Python\Python37\site-packages\swat\cas\rest\connection.py in __init__(self, hostname, port, username, password, soptions, error)
    359                             raise SWATError('%s (%s)' % (out['error'], out['details']))
--> 360                         raise SWATError(out['error'])
    361 

SWATError: HTTP/1.1 401 Unauthorized

During handling of the above exception, another exception occurred:

SWATError                                 Traceback (most recent call last)
<ipython-input-14-b9cc9b42f0b2> in <module>
      7 os.environ['CAS_TRACE_UI_ACTIONS']="TRUE"
      8 
----> 9 conn = swat.CAS(sas_server_url+'/'+cas_server_name+'-http/',username='', password=access_token,  protcol='https')

~\AppData\Roaming\Python\Python37\site-packages\swat\cas\connection.py in __init__(self, hostname, port, username, password, session, locale, nworkers, name, authinfo, protocol, path, ssl_ca_list, **kwargs)
    424                 params = (hostname, port, username, password, soptions, self._sw_error)
    425                 if protocol in ['http', 'https']:
--> 426                     self._sw_connection = rest.REST_CASConnection(*params)
    427                 else:
    428                     self._sw_connection = clib.SW_CASConnection(*params)

~\AppData\Roaming\Python\Python37\site-packages\swat\cas\rest\connection.py in __init__(self, hostname, port, username, password, soptions, error)
    380 
    381             except Exception as exc:
--> 382                 raise SWATError(str(exc))
    383 
    384     def _set_next_connection(self):

SWATError: HTTP/1.1 401 Unauthorized
```

If I pass the equivalent action to the CAS API directly, but instead use 'Authorization' : 'Bearer ' + token in the header I get can create a session through the CAS REST API directly:

Code:
```
import requests
import json
import pprint

# use existing oauth token to connect to CAS through CAS http
headers = {
    'Content-Type': 'application/json',
    'Content-Length': '0',
    'Authorization': 'Bearer {0}'.format(str(access_token)) 
}
s = requests.Session()
r = s.put(sas_server_url+'/cas-shared-default-http/cas/sessions', verify=False, headers = headers)
json_data = json.loads(r.text)

active_cas_session=json_data['session']
pprint.pprint(json_data)

```

Log:
```{'Authorization': 'Bearer '
'aklsjdflkasdjfkljsdlkfj',
 'Content-Length': '0',
 'Content-Type': 'application/json'}

'{ "session": "7025b47f-f57d-8f43-ba73-383cbc424c76" }\n'
```

I haven't been able to get a Basic authorization header to work with the OAuth token, including the base64 format I see used in the swat source code for authorization.

**Comment by kesmit13** [comment]
@GMoney1986 It looks like this was a bug (although I swear using Basic used to work). 

BTW, I have a couple of other code cleanup tips you may be interested in. You used environment variables above to set options. You can do this using the `swat.options` object as well, which may look a bit more natural in the Python code. I typically just use the env vars in the shell or containers.
```
swat.options.cas.debug.requests = True
swat.options.cas.debug.request_bodies = True
swat.options.cas.debug.responses = True
swat.options.cas.trace_ui_actions = True
```

And for the `hostname` parameter of the CAS constructor, you can use a URL there and drop the `protocol` parameter. If your `sas_server_url` variable has https in it already, you don't need the `protocol` parameter.  
```
swat.CAS('https://%s:443/cas-shared-default-http/' % sas_server_name, password=oauth_token)
```

This also works with the binary protocol.
```
swat.CAS('cas://%s:5570' % sas_server_name)
```

**Comment by GMoney1986** [comment]
Thank you for the tips and insight! I'll implement what you've suggested and dig a bit deeper on why basic connection is not working on my end. 

At worst, I can always butcher the source code to force the Bearer Authorization if I get truly desperate.

**Comment by kesmit13** [comment]
@GMoney1986 I did push a fix for the issue you found: 697671b. You'd have to install from the source on Github rather than using a release until a new release comes out though.
